Make fitting more robust and add parallelization

stefanhahmann opened the pull request:
This pull request fixes some issues when running the ellipsoid fitting:

* The fitting continues in cases where individual ellipsoid could not be fitted
* The fitting for multiple ellipsoid is run in parallel now
* Logic that may have resulted in endless loops is removed.

Commit:
Allow ellipsoid fitting to finish also in cases some ellipsoid could not be fitted at all

Commit:
Add parallelization to ellipsoid fitting and avoid endless loop while computing candidate ellipsoids

Commit:
Avoid endless loop in case less edgels than required points/edgels for ellipsoid fitting

Commit:
Update mastodon version number to be used for ellipsoid fitting plugin

Commit:
Add write lock while updating ellipsoid covariances

Commit:
Add todo regarding num samples parameter

Commit:
Comment out some log statements and now remove unused code to enforce a given number of samplings (was commented out, since i may have lead to endless loops)

maarzt commented on src/main/java/org/mastodon/mamut/fitting/edgel/SampleEllipsoidEdgel.java:
```diff
@@ -118,33 +129,40 @@ public static Ellipsoid sample(
 			final double outsideCutoffDistance,
 			final double insideCutoffDistance,
-			final double angleCutoffDistance )
+			final double angleCutoffDistance ) throws IllegalArgumentException
```
Please remove the unnecessary `throws IllegalArgumentException`. A javadoc that describes what the method does would probably be really helpful. Please describe under which condition the method is expected to throw the IllegalArgumentException.

stefanhahmann replied:
Good advice.
I removed throwing IllegalArgumentException in method fitToInliers (instead just the return the given ellipsoid in the case where that Exception would be thrown). I also added javadoc to the method SampleEllipsoidEdgel.fitToInliers(), however, I am unsure, if this javadoc correctly says what the method does. I would appreciate, if you could check.

maarzt replied:
It's ok now

maarzt commented on src/main/java/org/mastodon/mamut/fitting/FitEllipsoidPlugin.java:
```diff
@@ -168,7 +169,7 @@ private void fitSelectedVertices()
 		final int sourceIndex = 0;
 
-//		System.out.println( "fitSelectedVertices()" );
+		//		System.out.println( "fitSelectedVertices()" );
```
Please undo white space changes. This would make to PR easier to review.

stefanhahmann replied:
In the meantime, there are barely no changes left that are whitespace only.
It would be difficult to undo it.
Would it help reading the PR with this setting (hide whitespace)? 
![grafik](https://user-images.githubusercontent.com/10515534/225946209-309c7580-c2d9-4a7b-a1ad-f6190e98ca6f.png)

maarzt commented on src/main/java/org/mastodon/mamut/fitting/FitEllipsoidPlugin.java:
```diff
@@ -205,41 +206,52 @@ private < T extends RealType< T > > void process( final SourceAndConverter< T >
-//		if ( vertices.isEmpty() )
-//			System.err.println( "no vertex selected" );
+		//		if ( vertices.isEmpty() )
+		//			System.err.println( "no vertex selected" );
```
Writing a message to stdout and returning would be a good idea.

stefanhahmann replied:
I agree. I did it like this now.

maarzt commented on src/main/java/org/mastodon/mamut/fitting/FitEllipsoidPlugin.java:
```diff
@@ -205,41 +206,52 @@ private < T extends RealType< T > > void process( final SourceAndConverter< T >
+				long[] widthHeightDepth = longArrayFrom( cropped, Interval::dimensions );
+				//System.out.println( "widthHeightDepth: " + widthHeightDepth[ 0 ] + " " + widthHeightDepth[ 1 ] + " "
+				//		+ widthHeightDepth[ 2 ] + ", spot: " + spot.getLabel() );
+				final RandomAccessibleInterval< FloatType > img = ArrayImgs.floats( widthHeightDepth );
```
```suggestion
final RandomAccessibleInterval< FloatType > img = ArrayImgs.floats( cropped.dimensionsAsLongArray() );
```

stefanhahmann replied:
I accepted the suggestion in the new location where this code is called.

maarzt commented on src/main/java/org/mastodon/mamut/fitting/ellipsoid/FitEllipsoid.java:
```diff
@@ -83,6 +86,12 @@ public static Ellipsoid yuryPetrov( final double[][] points )
 
 		final double[][] DTD = new double[ 9 ][ 9 ];
 		LinAlgHelpers.multATB( d, d, DTD );
+		CholeskyDecomposition choleskyDecomposition = new CholeskyDecomposition( new Matrix( DTD ) );
+
+		// symmetric, positive, definite?
+		if ( !choleskyDecomposition.isSPD() )
+			// in this case, ellipsoid fitting is not possible
+			return null;
 		final Matrix V = new CholeskyDecomposition( new Matrix( DTD ) ).solve( new Matrix( b ) );
```
The code calls ```new CholeskyDecomposition( new Matrix( DTD ) )``` twice. Is there a reason for that. I think calling it once would be enough.

stefanhahmann replied:
You are completely right. I changed it.

maarzt commented on src/main/java/org/mastodon/mamut/fitting/FitEllipsoidPlugin.java:
```diff
@@ -187,59 +194,137 @@ private void fitSelectedVertices()
+		return a;
+	}
+
+	private class SpotEllipsoidFittingTask< T extends RealType< T > > implements Runnable
+	{
+		private final Spot spot;
+
+		private final SourceAndConverter< T > source;
+
+		private final AffineTransform3D sourceToGlobal;
+
+		private final MamutAppModel appModel;
+
+		private final int totalTasks;
+
+		public SpotEllipsoidFittingTask(
+				final Spot spot, final SourceAndConverter< T > source, final AffineTransform3D sourceToGlobal,
+				final MamutAppModel appModel, final int totalTasks )
+		{
+			this.spot = spot;
+			this.source = source;
+			this.sourceToGlobal = sourceToGlobal;
+			this.appModel = appModel;
+			this.totalTasks = totalTasks;
+		}
```
This class could be replaced by a method:
```java
private void fitEllipsoid(
		final Spot spot, final SourceAndConverter< T > source, final AffineTransform3D sourceToGlobal,
		final MamutAppModel appModel, final int totalTasks ) {
	// here should be all the content off the run method
}
```
which could be called via a lambda expression:
```java
todo.add( Executors.callable( () -> fitEllipsoid( spotCopy, source, sourceToGlobal, appModel, vertices.size() ) ) );
```
The result would be easier to read and understand.

stefanhahmann replied:
I disagree that this is easier to read and understand, since it hides the generation of the Task in the Lambda-expression "() -> ", which I do find not particularly easy to read. But unfortunately, I do not seem to be in the majority with this opinion, so I changed it anyway.

maarzt commented on src/main/java/org/mastodon/mamut/fitting/edgel/SampleEllipsoidEdgel.java:
```diff
@@ -96,11 +107,15 @@ public static Ellipsoid sample(
 				final Ellipsoid ellipsoid = FitEllipsoid.yuryPetrov( coordinates );
 
-				// don't count degenerate samples
+				// skip samples that cannot be fitted to an ellipsoid
```
The `FitEllipsoid.yuryPetrov(...)` method has several ways to fails:
1. return null
2. return an ellipsoid with radii that is NaN
3. throw IlleagalArgumentExcption
4. throw RuntimeException

This is not nice and very complex to handle. This PR doesn't introduce this problem, but very much tries to solve the consequences. You are understanding well under which conditions these errors occur, and how to handle them. Please describe it in javadoc or a comment.

stefanhahmann replied:
Good advice. I tried my best to add javadoc to the methods FitEllipsoid.yuryPetrov() and SampleEllipsoidEdgel.sample() describing parameters and potential reasons for failure, but I am really not sure, if I did it correctly. Would be happy, if you could check the results of that.

maarzt reviewed (commented):
Unfortunately the code is very hard to read an understand. This PR makes methods that are already too long, even longer. There are two things I would suggest:

1. Add a unit test that proofs that the code now works better than before.
2. Split the large methods into smaller methods. Try to better separate the code for debugging from the code that performs the actual calculations.

Commit:
Refactor SpotEllipsoidFittingTask to Lambda expression using fitEllipsoid()

Commit:
Do not call choleskyDecomposition unnecessarily twice

Commit:
Add javadoc to method SampleEllipsoidEdgel.sample

Commit:
Add javadoc to method FitEllipsoid.yuryPetrov

Commit:
Make method fitToInliers() private

Commit:
Make method yuryPetrov() return null if one the radii of the fitted ellipsoid is Double.NaN

Commit:
Add javadoc to method fitToInliers

Commit:
Method fitToInliers returns the original ellipsoid if fitting fails

Commit:
Move method getFittedEllipsoid to class FitEllipsoid

Commit:
Add a demo app that renders a single blog and an ellipsoid

Commit:
Return immediately from ellipsoid fitting, if no vertex is selected

Commit:
Replace image conversion method

Commit:
Move methods renderMultivariateNormalDistribution and scalarProduct to new class BlobRenderingUtils

Commit:
Move debug infos to new method showDebugInfos and introduce some class constants for this

Commit:
Add unit test for multiple ellipsoid fittings with slightly changed parameters

Commit:
Refactor method names in FitEllipsoidTest

Commit:
Introduce new class FitEllipsoidProcessor and according unit test

Commit:
Refactor methods in BlobRenderingUtils

Commit:
Add test showing that parallel processing is faster

Commit:
Add test showing more fitting to find the best ellipsoid leads to a higher rate of accepted ellipsoid fittings

Commit:
Relax acceptance criterion for ellipsoid axes length to 30% allowed relative difference

stefanhahmann commented:
> Unfortunately the code is very hard to read an understand. This PR makes methods that are already too long, even longer. There are two things I would suggest:
> 
>     1. Add a unit test that proofs that the code now works better than before.

I added a unit test showing that parallelization is quicker while achieving that the same amount of ellipsoid could be fitted.

I also added a unit test showing that increasing the number of fits to be used to find the best fitting ellipsoid leads to better results. However, the improvement is unfortunately not really significant.

> 
>     2. Split the large methods into smaller methods. Try to better separate the code for debugging from the code that performs the actual calculations.

I tried to separate code for calculations from code for debugging.


> Unfortunately the code is very hard to read an understand. This PR makes methods that are already too long, even longer. There are two things I would suggest:
> 
>     1. Add a unit test that proofs that the code now works better than before.
> 
>     2. Split the large methods into smaller methods. Try to better separate the code for debugging from the code that performs the actual calculations.

Commit:
Remove unrequired empty line

maarzt committed:
SampleEllipsoidEdgel: refactoring, simplify the code

maarzt committed:
FitEllipsoidProcessorTest: speed up the setup method

Cropping the image with Views.interval( image, interval ) before applying
BlobRenderingUtils.renderMultivarianteNormalDistribution( ... ) allows a
significant speedup.

maarzt committed:
FitEllipsoidPlugin: remove unused code

maarzt committed:
FitEllipsoidProcessorTest: make the covariance matrix is positive definite

maarzt committed:
FirEllipsoidProcessorTest: refactor the test

maarzt committed:
FitEllipsoidProcessor: use imglib2 Parallelization

maarzt commented on src/test/java/org/mastodon/mamut/fitting/ellipsoid/FitEllipsoidTest.java:
```diff
@@ -0,0 +1,195 @@
+		return true;
+	}
+
+	private static Triple< Double, Double, Double > getEigenvalues( double[][] matrix )
```
The test duplicates the code for computing ellipsoid axis, and thereby adds a wrongly named method getEigenvalues, that actually computes the square root of eigenvalues.

stefanhahmann replied:
Very true. It should have been named getAxes() and also the code duplication was not good.
Regarding the latter: I am wondering, where a method that computes ellipsoid axes from a covariance matrix could have actually gone within the Mastodon repos so that code duplication would not have been necessary?

tinevez replied:
Totally right.

maarzt reviewed (changes requested):
I would like to clearify why I would like to reject this PR:

A [big piece](https://github.com/mastodon-sc/mastodon-ellipsoid-fitting/blob/9d5183a527219ab8ad9adca54e15f51b77f51a32/src/main/java/org/mastodon/mamut/fitting/ellipsoid/FitEllipsoid.java#L191-L281) of very complicated code is moved to the FitEllipsoid class, into a package where it doesn't belong.
_(The entire package "org.mastodon.mamut.fitting.ellipsoid" does basic geometry on ellipsoids an points. It is independent of mastodon, only depends on a view imglib2 classes and Jama for eigenvalue computation. The code that was moved into the package in contrast depended on large central parts of Mastodon.  Classes like MamutAppModel, Spot and also BigDataViewer.
Moving the code there also introduced an circular dependency FitEllipsoid -> SampleEllipsoidEdgel -> FitEllipsoid between classes. Which made it hard to understand how one functions build onto each other.)_

The code for testing is long and complex but actually fails to show the the FitEllipsoidPlugin actually works. The results of the code running in parallel (FitEllipsoidProcessor) are never tested. Additionally the test tries to show that the parallel version is faster. But it compares it to a new single threaded version. Which is not a fair comparison.

I think it was me asking for tests, that caused you to further complicate the code. I'm sorry that I  failed to ask for changes in a way that allowed you to improve the code. I took the initial status of your PR, added a simpler test and did some refactoring. The new PR is here #4

maarzt commented:
Closed in favor of #4
